pkg/upgrade/upgrades/wait_for_del_range_in_gc_job.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package upgrades
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+// waitForDelRangeInGCJob ensures that any running GC jobs have adopted the new
+// DeleteRange protocol.
+func waitForDelRangeInGCJob(
+	ctx context.Context, _ clusterversion.ClusterVersion, deps upgrade.TenantDeps, _ *jobs.Job,
+) error {
+	for r := retry.StartWithCtx(ctx, retry.Options{
+		Closer: nil,
+	}); r.Next(); {
+		jobIDs, err := collectJobIDsFromQuery(
+			ctx, deps.InternalExecutor, "wait-for-gc-job-upgrades", `
+    WITH jobs AS (
+                SELECT *
+                  FROM (
+                        SELECT id,
+                               crdb_internal.pb_to_json('payload', payload) AS pl,
+                               crdb_internal.pb_to_json('progress', progress) AS progress
+                          FROM system.jobs
+                         WHERE status IN `+jobs.NonTerminalStatusTupleString+`
+                       )
+                 WHERE (pl->'schemaChangeGC') IS NOT NULL
+              ),
+         tables AS (
+                    SELECT id,
+                           json_array_elements(pl->'schemaChangeGC'->'tables') AS pl,
+                           json_array_elements(progress->'schemaChangeGC'->'tables') AS progress
+                      FROM jobs
+                ),
+         indexes AS (
+                    SELECT id,
+                           json_array_elements(pl->'schemaChangeGC'->'indexes'),
+                           json_array_elements(progress->'schemaChangeGC'->'indexes')
+                      FROM jobs
+                 ),
+         elements AS (SELECT * FROM tables UNION ALL SELECT * FROM indexes)
+  SELECT id
+    FROM elements
+   WHERE COALESCE(progress->>'status' NOT IN ('WAITING_FOR_MVCC_GC', 'CLEARED'), true)
+GROUP BY id;
+`)
+		if err != nil || len(jobIDs) == 0 {
+			return err
+		}
+		log.Infof(ctx, "waiting for %d GC jobs to adopt the new protocol: %v", len(jobIDs), jobIDs)
```

Comment:
If any of these jobs are in backoff due to errors, can we end up waiting for hours here? Or are they accounted for by `nonTerminalStatusTupleString`? If yes, would it be safe to ignore these since they would adopt the new protocol once they resume?

Reply:
This was a good suggestion! Thanks.